src/TypeScriptGenerator.ts

```diff
@@ -842,50 +843,64 @@ function selectionsToRawResponseBabel(
 
   if (Object.keys(byConcreteType).length) {
     const baseFieldsMap = selectionsToMap(baseFields);
-
     for (const concreteType in byConcreteType) {
+      const mergedSeletions = Array.from(
+        mergeSelections(
+          baseFieldsMap,
+          selectionsToMap(byConcreteType[concreteType]),
+          false
+        ).values()
+      );
       types.push(
-        Array.from(
-          mergeSelections(
-            baseFieldsMap,
-            selectionsToMap(byConcreteType[concreteType]),
-            false
-          ).values()
-        ).map(selection => {
-          if (isTypenameSelection(selection)) {
-            return makeRawResponseProp(
-              schema,
-              { ...selection, conditional: false },
-              state,
-              concreteType
-            );
-          }
-
-          return makeRawResponseProp(schema, selection, state, concreteType);
-        })
+        exactObjectTypeAnnotation(
+          // @ts-ignore
+          mergedSeletions.map(selection =>
+            makeRawResponseProp(schema, selection, state, concreteType)
+          )
+        ) as any
       );
+      appendLocal3DPayload(types, mergedSeletions, schema, state, concreteType);
     }
   }
-  if (baseFields.length) {
+  if (baseFields.length > 0) {
     types.push(
-      baseFields.map(selection => {
-        if (isTypenameSelection(selection)) {
-          return makeRawResponseProp(
-            schema,
-            { ...selection, conditional: false },
-            state,
-            nodeTypeName
-          );
-        }
-
-        return makeRawResponseProp(schema, selection, state, null);
-      })
+      exactObjectTypeAnnotation(
+        // @ts-ignore
+        baseFields.map(selection =>
+          makeRawResponseProp(schema, selection, state, nodeTypeName)
+        )
+      ) as any
     );
+    appendLocal3DPayload(types, baseFields, schema, state, nodeTypeName);
   }
+  return ts.createUnionTypeNode(types as any);
+}
 
-  return ts.createUnionTypeNode(
-    types.map(props => exactObjectTypeAnnotation(props))
-  );
+function appendLocal3DPayload(
+  types: any[],
+  selections: ReadonlyArray<Selection>,
+  schema: Schema,
+  state: State,
+  currentType?: string | null
+): void {
+  const moduleImport = selections.find(sel => sel.kind === "ModuleImport");
+  if (moduleImport) {
+    // Generate an extra opaque type for client 3D fields
+    state.runtimeImports.add("Local3DPayload");
+    types.push(
+      ts.createTypeReferenceNode(ts.createIdentifier("Local3DPayload"), [
+        stringLiteralTypeAnnotation(moduleImport.documentName as any),
+        exactObjectTypeAnnotation(
+          // @ts-ignore
```

Comment:
Why?

Reply:
@alloy The problem is that exactObjectTypeAnnotation arg is `ts.PropertySignature[]`, but `selections` is `ReadonlyArray<Selection>`.
How could I fix this?